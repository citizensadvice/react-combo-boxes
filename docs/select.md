# `<Select>`

Generates an HTML `<select>`, but with a more convenient way of setting the options and responding to changed values.

This uses the same basic props as `<DropDown>` and `<ComboBox>` and they maybe used somewhat interchangeably.

> :warning: Some users find `<select>` controls difficult to use.  A radio group may be more appropriate.

## Usage

This is a stateless controlled component.  You must respond to `onValue` or `onChange` to update the selected value.

```js
const [value, setValue] = useState(initialValue);

<Select
  options={options}
  value={value}
  onValue={setValue}
/>
```

| Prop            | Type       | Purpose                                   |
| ----            | ----       | ----                                      |
| `placeholder`   | `String`   | Set a placeholder option                  |
| `options`       | `Array`    | The set of options. [See options][1]      |
| `value`         | `Any`      | The currently selected option             |
| `onValue`       | `Function` | Called with the selected option           |
| `mapOption`     | `Function` | Use to map options. [See options][1]      |
| `optionProps`   | `Object`   | Props to add to all `<option>` elements   |
| `optGroupProps` | `Object`   | Props to add to all `<optgroup>` elements |
| `ref`           | `ref`      | Will be forwarded to `<select>` element   |
| Any other prop  |            | Will be passed to the `<select>` element  |

[1]: options.md
