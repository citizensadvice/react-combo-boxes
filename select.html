<html lang="en-GB"><head><meta charset="utf-8"><title>Select</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" href="/react-combo-boxes/styles.97e0bef6.css"><script src="https://polyfill.io/v3/polyfill.min.js?features=default%2Ces2018%2Ces2015%2Ces2016%2Ces2017%2Ces2019%2CNodeList.prototype.forEach" crossorigin="anonymous"></script></head><body><nav><div>React combo boxes</div><details><summary>Select</summary><a>About</a><a href="/react-combo-boxes/select/string_options/index.html">String options</a><a href="/react-combo-boxes/select/object_options/index.html">Object options</a><a href="/react-combo-boxes/select/placeholder_option/index.html">Placeholder option</a><a href="/react-combo-boxes/select/disabled_option/index.html">Disabled option</a><a href="/react-combo-boxes/select/grouped_options/index.html">Grouped options</a><a href="/react-combo-boxes/select/mapped_options/index.html">Mapped options</a><a href="/react-combo-boxes/select/html_options/index.html">HTML options</a></details><details><summary>Drop down</summary><a href="/react-combo-boxes/drop_down.html">About</a><a href="/react-combo-boxes/drop_down/string_options/index.html">String options</a><a href="/react-combo-boxes/drop_down/object_options/index.html">Object options</a><a href="/react-combo-boxes/drop_down/placeholder_option/index.html">Placeholder option</a><a href="/react-combo-boxes/drop_down/disabled_option/index.html">Disabled option</a><a href="/react-combo-boxes/drop_down/disabled/index.html">Disabled drop down</a><a href="/react-combo-boxes/drop_down/grouped_options/index.html">Grouped options</a><a href="/react-combo-boxes/drop_down/html_options/index.html">HTML options</a><a href="/react-combo-boxes/drop_down/mapped_options/index.html">Mapped options</a><a href="/react-combo-boxes/drop_down/without_managed_focus/index.html">Without managed focus</a><a href="/react-combo-boxes/drop_down/aria_1_2_groups/index.html">With ARIA 1.2 groups</a><details><summary>Edge cases</summary><a href="/react-combo-boxes/drop_down/identical_options/index.html">Identical options</a><a href="/react-combo-boxes/drop_down/no_options/index.html">No options</a><a href="/react-combo-boxes/drop_down/all_disabled_options/index.html">All disabled options</a></details></details><details><summary>Combo box</summary><a href="/react-combo-boxes/combo_box.html">About</a><a href="/react-combo-boxes/combo_box/string_options/index.html">String options</a><a href="/react-combo-boxes/combo_box/object_options/index.html">Object options</a><a href="/react-combo-boxes/combo_box/disabled_option/index.html">Disabled option</a><a href="/react-combo-boxes/combo_box/grouped_options/index.html">Grouped options</a><a href="/react-combo-boxes/combo_box/html_options/index.html">HTML options</a><a href="/react-combo-boxes/combo_box/mapped_options/index.html">Mapped options</a><a href="/react-combo-boxes/combo_box/without_autocomplete/index.html">Without autocomplete</a><a href="/react-combo-boxes/combo_box/without_managed_focus/index.html">Without managed focus</a><a href="/react-combo-boxes/combo_box/inline_autocomplete/index.html">With inline autocomplete</a><a href="/react-combo-boxes/combo_box/aria_1_0/index.html">ARIA 1.0</a><a href="/react-combo-boxes/combo_box/aria_1_1/index.html">ARIA 1.1</a><a href="/react-combo-boxes/combo_box/aria_1_2_groups/index.html">ARIA 1.2 groups</a><a href="/react-combo-boxes/combo_box/listbox_footer/index.html">Listbox footer</a><details><summary>Searches</summary><a href="/react-combo-boxes/searchers.html">About</a><a href="/react-combo-boxes/combo_box/async_search/index.html">Async search</a><a href="/react-combo-boxes/combo_box/search_suggestions/index.html">Search suggestions</a><a href="/react-combo-boxes/combo_box/background_autocomplete/index.html">With background autocomplete</a></details><details><summary>Edge cases</summary><a href="/react-combo-boxes/combo_box/identical_options/index.html">Identical options</a><a href="/react-combo-boxes/combo_box/all_disabled_options/index.html">All disabled options</a><a href="/react-combo-boxes/combo_box/autoselect_disabled_option/index.html">Disabled option and autoselect</a><a href="/react-combo-boxes/combo_box/search_min_length/index.html">Search min length</a></details></details><details><summary>Table layout</summary><a href="/react-combo-boxes/list_box_table.html">About</a><details><summary>Drop down</summary><a href="/react-combo-boxes/list_box_table/drop_down/index.html">Table</a><a href="/react-combo-boxes/list_box_table/drop_down_headers/index.html">With headers</a><a href="/react-combo-boxes/list_box_table/drop_down_grouped/index.html">Grouped</a></details><details><summary>Combo box</summary><a href="/react-combo-boxes/list_box_table/combo_box/index.html">Table</a><a href="/react-combo-boxes/list_box_table/combo_box_headers/index.html">With headers</a><a href="/react-combo-boxes/list_box_table/combo_box_grouped/index.html">Grouped</a></details></details><a href="/react-combo-boxes/test_helper.html">Test helper</a></nav><script src="/react-combo-boxes/side_collapse.da5ebe24.js"></script><main><h1 id="select"><code>&lt;Select&gt;</code></h1>
<p>Generates an HTML <code>&lt;select&gt;</code>, but with a more convenient way of setting the options and responding to changed values.</p>
<p>This uses the same basic props as <code>&lt;DropDown&gt;</code> and <code>&lt;ComboBox&gt;</code> and they maybe used somewhat interchangeably.</p>
<blockquote>
<p>⚠️ Some users find <code>&lt;select&gt;</code> controls difficult to use. A radio group may be more appropriate.</p>
</blockquote>
<h2 id="usage">Usage</h2>
<p>This is a stateless controlled component. You must respond to <code>onValue</code> or <code>onChange</code> to update the selected value.</p>
<pre><code class="language-js"><span class="token keyword">const</span> <span class="token punctuation">[</span>value<span class="token punctuation">,</span> setValue<span class="token punctuation">]</span> <span class="token operator">=</span> <span class="token function">useState</span><span class="token punctuation">(</span>initialValue<span class="token punctuation">)</span><span class="token punctuation">;</span>

<span class="token operator">&lt;</span>Select
  options<span class="token operator">=</span><span class="token punctuation">{</span>options<span class="token punctuation">}</span>
  value<span class="token operator">=</span><span class="token punctuation">{</span>value<span class="token punctuation">}</span>
  onValue<span class="token operator">=</span><span class="token punctuation">{</span>setValue<span class="token punctuation">}</span>
<span class="token operator">/</span><span class="token operator">></span></code></pre>
<table>
<thead>
<tr>
<th>Prop</th>
<th>Type</th>
<th>Purpose</th>
</tr>
</thead>
<tbody><tr>
<td><code>placeholder</code></td>
<td><code>String</code></td>
<td>Set a placeholder option</td>
</tr>
<tr>
<td><code>options</code></td>
<td><code>Array</code></td>
<td>The set of options. <a href="/react-combo-boxes/options.html">See options</a></td>
</tr>
<tr>
<td><code>value</code></td>
<td><code>Any</code></td>
<td>The currently selected option</td>
</tr>
<tr>
<td><code>onValue</code></td>
<td><code>Function</code></td>
<td>Called with the selected option</td>
</tr>
<tr>
<td><code>mapOption</code></td>
<td><code>Function</code></td>
<td>Use to map options. <a href="/react-combo-boxes/options.html">See options</a></td>
</tr>
<tr>
<td><code>optionProps</code></td>
<td><code>Object</code></td>
<td>Props to add to all <code>&lt;option&gt;</code> elements</td>
</tr>
<tr>
<td><code>optGroupProps</code></td>
<td><code>Object</code></td>
<td>Props to add to all <code>&lt;optgroup&gt;</code> elements</td>
</tr>
<tr>
<td><code>ref</code></td>
<td><code>ref</code></td>
<td>Will be forwarded to <code>&lt;select&gt;</code> element</td>
</tr>
<tr>
<td>Any other prop</td>
<td></td>
<td>Will be passed to the <code>&lt;select&gt;</code> element</td>
</tr>
</tbody></table>
</main><script src="/react-combo-boxes/render.c763a54f.js"></script></body></html>