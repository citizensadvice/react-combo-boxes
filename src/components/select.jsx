import React, { useCallback, forwardRef } from 'react';
import PropTypes from 'prop-types';
import { useNormalisedOptions } from '../hooks/use_normalised_options';
import { renderGroupedOptions } from '../helpers/render_grouped_options';

export const Select = forwardRef((rawProps, ref) => {
  const {
    options, onChange, onValue, value: _, selectedOption,
    optionProps, optgroupProps,
    ...props
  } = useNormalisedOptions(rawProps, { mustHaveSelection: true });

  const handleChange = useCallback((e) => {
    onValue?.(options.find((o) => o.identity === e.target.value)?.value);
    onChange?.(e);
  }, [onValue, onChange, options]);

  return (
    <select
      value={selectedOption?.identity ?? ''}
      onChange={handleChange}
      ref={ref}
      {...props}
    >
      {renderGroupedOptions({
        options,
        renderGroup({ key, html, children, label }) { // eslint-disable-line react/prop-types
          return (
            <optgroup
              key={key}
              label={label}
              {...optgroupProps}
              {...html}
            >
              {children}
            </optgroup>
          );
        },
        // eslint-disable-next-line react/prop-types
        renderOption({ identity, label, key, html, disabled }) {
          return (
            <option
              value={identity}
              key={key}
              disabled={disabled}
              {...optionProps}
              {...html}
            >
              {label}
            </option>
          );
        },
      })}
    </select>
  );
});

Select.propTypes = {
  placeholder: PropTypes.node,
  onChange: PropTypes.func,
  onValue: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.any).isRequired,
  value: PropTypes.any,
  optionProps: PropTypes.object,
  optgroupProps: PropTypes.object,
};

Select.defaultProps = {
  placeholder: null,
  value: null,
  onChange: null,
  onValue: null,
  optionProps: null,
  optgroupProps: null,
};

Select.displayName = 'Select';
