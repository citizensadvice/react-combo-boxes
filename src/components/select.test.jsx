import React, { useState, forwardRef } from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Select } from './select';

const SelectWrapper = forwardRef(({ value: initialValue, ...props }, ref) => {
  const [value, setValue] = useState(initialValue);
  return (
    <Select value={value} onValue={setValue} {...props} ref={ref} />
  );
});

describe('options', () => {
  describe('as array of strings', () => {
    const options = ['Apple', 'Banana', 'Orange'];

    it('renders an select', () => {
      const { container } = render(
        <Select options={options} />,
      );
      expect(container).toMatchSnapshot();
    });

    it('renders a select with a selected value', () => {
      const { getByRole } = render(
        <SelectWrapper options={options} value="Orange" />,
      );
      expect(getByRole('combobox')).toHaveValue('Orange');
    });

    it('triggers the onValue callback with the selected value', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: 'Banana' } });
      expect(spy).toHaveBeenCalledWith('Banana');
    });

    it('triggers the onChange callback with the selected value', () => {
      const spy = jest.fn((e) => e.persist());
      const { getByRole } = render(<SelectWrapper options={options} onChange={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: 'Banana' } });
      expect(spy).toHaveBeenCalledWith(expect.objectContaining({
        target: expect.objectContaining({
          nodeName: 'SELECT',
          value: 'Banana',
        }),
      }));
    });

    it('updates when the value changes', () => {
      const { getByRole } = render(<SelectWrapper options={options} />);
      fireEvent.change(getByRole('combobox'), { target: { value: 'Apple' } });
      expect(getByRole('combobox')).toHaveValue('Apple');
    });

    it('triggers onValue when selecting an empty string', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper options={['foo', '']} onValue={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(spy).toHaveBeenCalledWith('');
    });
  });

  describe('options as array of numbers', () => {
    const options = [0, 1, 2, 3];

    it('renders an select', () => {
      const { container } = render(
        <SelectWrapper options={options} />,
      );
      expect(container).toMatchSnapshot();
    });

    it('triggers the onValue callback with the selected value', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '2' } });
      expect(spy).toHaveBeenCalledWith(2);
    });

    it('triggers the onChange callback with the selected value', () => {
      const spy = jest.fn((e) => e.persist());
      const { getByRole } = render(<SelectWrapper options={options} onChange={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '2' } });
      expect(spy).toHaveBeenCalledWith(expect.objectContaining({
        target: expect.objectContaining({
          nodeName: 'SELECT',
          value: '2',
        }),
      }));
    });

    it('updates when the value changes', () => {
      const { getByRole } = render(<SelectWrapper options={options} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '1' } });
      expect(getByRole('combobox')).toHaveValue('1');
    });

    it('triggers onValue when selecting 0', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '0' } });
      expect(spy).toHaveBeenCalledWith(0);
    });
  });

  describe('options as null', () => {
    const options = ['foo', null];

    it('renders an select', () => {
      const { container } = render(
        <SelectWrapper options={options} />,
      );
      expect(container).toMatchSnapshot();
    });

    it('triggers the onValue callback with the selected value', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(spy).toHaveBeenCalledWith(null);
    });

    it('triggers the onChange callback with the selected value', () => {
      const spy = jest.fn((e) => e.persist());
      const { getByRole } = render(<SelectWrapper options={options} value="foo" onChange={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(spy).toHaveBeenCalledWith(expect.objectContaining({
        target: expect.objectContaining({
          nodeName: 'SELECT',
          value: '',
        }),
      }));
    });

    it('updates when the value changes', () => {
      const { getByRole } = render(<SelectWrapper options={options} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(getByRole('combobox')).toHaveValue('');
    });
  });

  describe('options as undefined', () => {
    const options = ['foo', undefined];

    it('renders an select', () => {
      const { container } = render(
        <SelectWrapper options={options} />,
      );
      expect(container).toMatchSnapshot();
    });

    it('triggers the onValue callback with the selected value', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(spy).toHaveBeenCalledWith(undefined);
    });

    it('triggers the onChange callback with the selected value', () => {
      const spy = jest.fn((e) => e.persist());
      const { getByRole } = render(<SelectWrapper options={options} onChange={spy} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(spy).toHaveBeenCalledWith(expect.objectContaining({
        target: expect.objectContaining({
          nodeName: 'SELECT',
          value: '',
        }),
      }));
    });

    it('updates when the value changes', () => {
      const { getByRole } = render(<SelectWrapper options={options} />);
      fireEvent.change(getByRole('combobox'), { target: { value: '' } });
      expect(getByRole('combobox')).toHaveValue('');
    });
  });

  describe('options as array of objects', () => {
    describe('label', () => {
      const options = [{ label: 'Apple' }, { label: 'Banana' }, { label: 'Orange' }];

      it('renders an select', () => {
        const { container } = render(
          <SelectWrapper options={options} />,
        );
        expect(container).toMatchSnapshot();
      });

      it('triggers the onValue callback with the selected value', () => {
        const spy = jest.fn();
        const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
        fireEvent.change(getByRole('combobox'), { target: { value: 'Banana' } });
        expect(spy).toHaveBeenCalledWith({ label: 'Banana' });
      });

      it('triggers the onChange callback with the selected value', () => {
        const spy = jest.fn((e) => e.persist());
        const { getByRole } = render(<SelectWrapper options={options} onChange={spy} />);
        fireEvent.change(getByRole('combobox'), { target: { value: 'Banana' } });
        expect(spy).toHaveBeenCalledWith(expect.objectContaining({
          target: expect.objectContaining({
            nodeName: 'SELECT',
            value: 'Banana',
          }),
        }));
      });

      it('updates when the value changes', () => {
        const { getByRole } = render(<SelectWrapper options={options} />);
        fireEvent.change(getByRole('combobox'), { target: { value: 'Banana' } });
        expect(getByRole('combobox')).toHaveValue('Banana');
      });
    });

    describe('value', () => {
      it('is used as a options identity', () => {
        const options = [{ label: 'foo', value: 1 }, { label: 'foo', value: 2 }, { label: 'foo', value: 3 }];
        const spy = jest.fn();
        const { getByRole } = render(<SelectWrapper options={options} value={2} onValue={spy} />);
        expect(getByRole('combobox')).toHaveValue('2');
        fireEvent.change(getByRole('combobox'), { target: { value: '3' } });
        expect(spy).toHaveBeenCalledWith({ label: 'foo', value: 3 });
      });
    });

    describe('id', () => {
      it('is used as a options identity', () => {
        const options = [{ label: 'foo', id: 1 }, { label: 'foo', id: 2 }, { label: 'foo', id: 3 }];
        const spy = jest.fn();
        const { getByRole } = render(<SelectWrapper options={options} value={2} onValue={spy} />);
        expect(getByRole('combobox')).toHaveValue('2');
        fireEvent.change(getByRole('combobox'), { target: { value: '3' } });
        expect(spy).toHaveBeenCalledWith({ label: 'foo', id: 3 });
      });
    });

    describe('disabled', () => {
      it('sets the disabled attribute', () => {
        const options = [{ label: 'foo', disabled: true }];
        const { getAllByRole } = render(<SelectWrapper options={options} />);
        expect(getAllByRole('option')[0]).toHaveAttribute('disabled', '');
      });
    });

    describe('html', () => {
      it('sets attributes on the element', () => {
        const options = [{ label: 'foo', html: { 'data-foo': 'bar', className: 'class' } }];
        const { getByRole } = render(<SelectWrapper options={options} />);
        expect(getByRole('option')).toHaveAttribute('data-foo', 'bar');
        expect(getByRole('option')).toHaveAttribute('class', 'class');
      });
    });

    describe('group', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Orange', group: 'Citrus' },
        { label: 'Lemon', group: 'Citrus' },
        { label: 'Raspberry', group: 'Berry' },
        { label: 'Strawberry', group: 'Berry' },
      ];

      it('renders grouped options', () => {
        const { container } = render(<SelectWrapper options={options} />);
        expect(container).toMatchSnapshot();
      });

      it('triggers the onValue callback with the selected value', () => {
        const spy = jest.fn();
        const { getByRole } = render(<SelectWrapper options={options} onValue={spy} />);
        fireEvent.change(getByRole('combobox'), { target: { value: 'Lemon' } });
        expect(spy).toHaveBeenCalledWith({ label: 'Lemon', group: 'Citrus' });
      });

      it('triggers the onChange callback with the selected value', () => {
        const spy = jest.fn((e) => e.persist());
        const { getByRole } = render(<SelectWrapper options={options} onChange={spy} />);
        fireEvent.change(getByRole('combobox'), { target: { value: 'Lemon' } });
        expect(spy).toHaveBeenCalledWith(expect.objectContaining({
          target: expect.objectContaining({
            nodeName: 'SELECT',
            value: 'Lemon',
          }),
        }));
      });

      it('updates when the value changes', () => {
        const { getByRole } = render(<SelectWrapper options={options} />);
        fireEvent.change(getByRole('combobox'), { target: { value: 'Lemon' } });
        expect(getByRole('combobox')).toHaveValue('Lemon');
      });
    });

    describe('other attributes', () => {
      it('does not render them', () => {
        const options = [{ label: 'foo', 'data-foo': 'bar' }];
        const { getByRole } = render(<SelectWrapper options={options} />);
        expect(getByRole('combobox')).not.toHaveAttribute('data-foo', 'bar');
      });
    });
  });

  describe('mapOption', () => {
    const options = [{ name: 'Apple' }, { name: 'Banana' }, { name: 'Orange' }];

    it('maps options', () => {
      const spy = jest.fn();
      const { getByRole } = render(<SelectWrapper
        options={options}
        onValue={spy}
        mapOption={({ name }) => ({ label: name })}
      />);
      fireEvent.change(getByRole('combobox'), { target: { value: 'Orange' } });
      expect(spy).toHaveBeenCalledWith({ name: 'Orange' });
    });

    it('selects a mapped option', () => {
      const { getByRole } = render(<SelectWrapper
        options={options}
        value={{ name: 'Banana' }}
        mapOption={({ name }) => ({ label: name })}
      />);
      expect(getByRole('combobox')).toHaveValue('Banana');
      fireEvent.change(getByRole('combobox'), { target: { value: 'Orange' } });
      expect(getByRole('combobox')).toHaveValue('Orange');
    });
  });
});

describe('placeholder', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('renders a placeholder option', () => {
    const { getByRole, getAllByRole } = render(<SelectWrapper options={options} placeholder="Please select…" />);
    expect(getAllByRole('option')[0]).toHaveTextContent('Please select…');
    expect(getByRole('combobox')).toHaveValue('');
  });

  it('renders with a selected value', () => {
    const { getByRole } = render(<SelectWrapper options={options} placeholder="Please select…" value="Orange" />);
    expect(getByRole('combobox')).toHaveValue('Orange');
  });
});

describe('optionProps', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('allows additional props to be added to all options', () => {
    const { getAllByRole } = render(
      <SelectWrapper options={options} optionProps={{ 'data-foo': 'bar' }} />,
    );
    getAllByRole('option').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe('optgroupProps', () => {
  const options = [
    { label: 'Apple' },
    { label: 'Orange', group: 'Citrus' },
    { label: 'Lemon', group: 'Citrus' },
    { label: 'Raspberry', group: 'Berry' },
    { label: 'Strawberry', group: 'Berry' },
  ];

  it('allows additional props to be added to all options', () => {
    const { container } = render(
      <SelectWrapper options={options} optgroupProps={{ 'data-foo': 'bar' }} />,
    );
    container.querySelectorAll('optgroup').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe('additional props', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('includes them on the select', () => {
    const { getByRole } = render(
      <SelectWrapper options={options} required data-foo="bar" />,
    );
    expect(getByRole('combobox')).toHaveAttribute('required', '');
    expect(getByRole('combobox')).toHaveAttribute('data-foo', 'bar');
  });
});

describe('ref', () => {
  it('references the select for an object ref', () => {
    const ref = { current: null };
    const { getByRole } = render((
      <SelectWrapper options={['foo']} ref={ref} />
    ));
    expect(ref.current).toEqual(getByRole('combobox'));
  });

  it('references the combobox for a function ref', () => {
    let value;
    const ref = (node) => {
      value = node;
    };
    const { getByRole } = render((
      <SelectWrapper options={['foo']} ref={ref} />
    ));
    expect(value).toEqual(getByRole('combobox'));
  });
});
