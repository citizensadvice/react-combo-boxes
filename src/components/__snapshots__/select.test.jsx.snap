// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options as array of strings renders an select 1`] = `
<div>
  <select>
    <option
      value="Apple"
    >
      Apple
    </option>
    <option
      value="Banana"
    >
      Banana
    </option>
    <option
      value="Orange"
    >
      Orange
    </option>
  </select>
</div>
`;

exports[`options options as array of numbers renders an select 1`] = `
<div>
  <select>
    <option
      value="0"
    >
      0
    </option>
    <option
      value="1"
    >
      1
    </option>
    <option
      value="2"
    >
      2
    </option>
    <option
      value="3"
    >
      3
    </option>
  </select>
</div>
`;

exports[`options options as array of objects group renders grouped options 1`] = `
<div>
  <select>
    <option
      value="Apple"
    >
      Apple
    </option>
    <optgroup
      label="Citrus"
    >
      <option
        value="Orange"
      >
        Orange
      </option>
      <option
        value="Lemon"
      >
        Lemon
      </option>
    </optgroup>
    <optgroup
      label="Berry"
    >
      <option
        value="Raspberry"
      >
        Raspberry
      </option>
      <option
        value="Strawberry"
      >
        Strawberry
      </option>
    </optgroup>
  </select>
</div>
`;

exports[`options options as array of objects label renders an select 1`] = `
<div>
  <select>
    <option
      value="Apple"
    >
      Apple
    </option>
    <option
      value="Banana"
    >
      Banana
    </option>
    <option
      value="Orange"
    >
      Orange
    </option>
  </select>
</div>
`;

exports[`options options as null renders an select 1`] = `
<div>
  <select>
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value=""
    />
  </select>
</div>
`;

exports[`options options as undefined renders an select 1`] = `
<div>
  <select>
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value=""
    />
  </select>
</div>
`;
